Kernel/SanCov.cpp

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (c) 2021, Patrick Meyer <git@the-space.agency>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Devices/KCOVDevice.h>
+#include <Kernel/Devices/KCOVInstance.h>
+#include <Kernel/FileSystem/FileDescription.h>
+#include <LibC/sys/ioctl_numbers.h>
+
+#include <Kernel/Panic.h>
+
+extern bool g_in_early_boot;
+
+extern "C" {
+
+void __sanitizer_cov_trace_pc(void);
+void __sanitizer_cov_trace_pc(void)
+{
+    if (g_in_early_boot) [[unlikely]]
+        return;
+
+    if (Processor::current().in_irq()) [[unlikely]] {
+        // Do not trace in interrupts.
+        return;
+    }
+
+    auto thread = Thread::current();
+    auto tid = thread->tid();
+    auto maybe_kcov_instance = KCOVDevice::thread_instance->get(tid);
```

Comment:
Is this safe without a lock?

Reply:
I believe so:
- We only ever trace 0 or 1 thread of any given process at a time. Additionally this function is injected into and called from the traced thread itself. Hence we can't have two threads writing to the buffer of any given KCOVInstance at the same time. 
- Furthermore KCOVInstance and its buffer only go away when the corresponding process dies. Hence we couldn't be here with an invalid buffer in KCOVInstance.

Does that make sense?